runtime.md

```diff
@@ -33,24 +33,79 @@ This is provided so that consumers can find the container's configuration and ro
 }
 ```
 
-## Lifecycle
+## Operations
+
+OCI compliant runtimes MUST support the following operations, unless support for the operation can not be supported by the base operating system (kernel) itself.
+
+OPEN: we should consider specifying the options/args for each operation.
+For example, checkpoint (for runc) seems to have a lot of params.
 
 ### Create
 
-Creates the container: file system, namespaces, cgroups, capabilities.
+Using the `config.json` and `runtime.json` files, along with the file system path referenced in the `config.json`, this operation MUST create a new container.
+This includes creating the relevant namespaces, cgroups and configuring the appropriate capabilities for the container.
+
+### Start
+
+This operation MUST start a previously created container.
+Starting a container MUST create a new process, as specified by the `config.json` file within the scope of the container.
+Attempting to start an already running container MUST have no effect on the container and MUST generate an error.
+Starting a stopped container, one that had been previously be running, MUST create a new process, as specified by the `config.json` file, within the scope of the container.
+
+OPEN: the entire "generate an error" stuff in here is something we should discuss.
+My initial thought was that it would be better to generate an error so that the call would know exactly what happened - meaning, did something new take place or was the container already running. If we didn't return an error (or something) then the caller wouldn't know.  And they can always ignore errors if they want/need to.
+
+### Stop
+
+This operation MUST stop a running container.
+Stopping a container MUST stop all of the processes running within the scope of the container.
+Stopping a container MUST NOT delete the associated namespaces or cgroups for the container.
+Attempting to stop a container that is not running MUST have no effect on the container and MUST generate an error.
+
+### Delete
+
+This operation MUST delete a container.
+Deleting a container MUST delete the associated namespaces and cgroups for the container.
+Deleting a container that is not stopped MUST first stop the container as specified by the `stop` operation.
+If the implied `stop` operations fails then the `delete` operation MUST fail.
+
+### Pause
+
+This operation MUST suspend all processes that are running within the container.
+If the container is not running, either stopped or aleady paused, then this operation MUST have no effect on the container and MUST generate an error.
+
+### Unpause
+
+This operation MUST resume all processes that were previously paused via the `pause` operation.
+If the container is not paused then this operation MUST have no effect on the container and MUST generate an error.
```

Comment:
On Wed, Oct 14, 2015 at 08:09:09PM -0700, Brandon Philips wrote:
> Is freezer idempotent across random tasks that join the namespace
> later? It seems like it operates per pid so I am not sure how this
> works if something just joins the namespaces.

I'm not sure what namespace you're talking about here, but it is
random processes that join the freezer cgroup later are frozen on
join:

  term1# mkdir /sys/fs/cgroup/freezer/abc

  term2# echo $$ >/sys/fs/cgroup/freezer/abc/cgroup.procs

  term1# echo FROZEN >/sys/fs/cgroup/freezer/abc/freezer.state
  term1# cat /sys/fs/cgroup/freezer/abc/freezer.state 
  FROZEN

  term3# echo $$ >/sys/fs/cgroup/freezer/abc/cgroup.procs

  term1# cat /sys/fs/cgroup/freezer/abc/cgroup.procs 
  15929
  16039
  term1# cat /sys/fs/cgroup/freezer/abc/freezer.state 
  FROZEN
  # ps aux | grep '15929\|16039' | grep -v grep
  root     15929  0.0  0.0  20788  4568 pts/11   D+   20:40   0:00 bash
  root     16039  0.0  0.0  20656  4352 pts/12   D+   20:43   0:00 bash